Add grun script for grammar testing

zklaus opened the pull request:
<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->
Closes #28.

<!--
Please add any other relevant info below:
-->
This adds a `grun` script the same way the package already provides a `antlr4` script. As the second common antlr command (see eg antlr's [Getting Started](https://github.com/antlr/antlr4/blob/master/doc/getting-started.md)), this helps with development and testing of grammars.

Commit:
Add grun script for grammar testing

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

Commit:
Bump build number

Commit:
Add Windows version of grun script

zklaus commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 3.20.2, conda-smithy 3.7.10, and conda-forge-pinning 2020.09.16.18.10.04

Commit:
Update github checksum

Unfortunately, checksums for tarballs in the github archive are not stable.
It seems that the checksum for this tarball has changed.
See also eg. https://github.com/easybuilders/easybuild-easyconfigs/issues/5151

bollwyvl commented on recipe/bld.bat:
```diff
@@ -11,5 +11,9 @@ cmd.exe /c mvn --batch-mode --projects=tool clean || echo ""
 echo IF %%ERRORLEVEL%% NEQ 0 EXIT /B %%ERRORLEVEL%% >> %LIBRARY_BIN%\antlr4.cmd
+
+echo java -Xmx500M -cp %LIBRARY_LIB%\\antlr4-%PKG_VERSION%-complete.jar;%%CLASSPATH%% org.antlr.v4.gui.TestRig %%* > %LIBRARY_BIN%\antlr4.cmd
+echo IF %%ERRORLEVEL%% NEQ 0 EXIT /B %%ERRORLEVEL%% >> %LIBRARY_BIN%\antlr4.cmd
```
looks like this is overwriting the `antlr4.cmd`... perhaps needs to be `grun.cmd`?

zklaus replied:
Good catch, thanks, fixed.

bollwyvl commented:
Thanks for looking into this!

Please also add a `grun` test to the `commands`.... a `--help` or `--version` would be fine, just to make sure it's working.

Commit:
Fix faulty Windows bld script

Commit:
Add grun to test commands

zklaus commented:
@bollwyvl, thanks for the comments! There is a problem with the maven build process. Do you have any idea on how to address that?

bollwyvl commented:
Eeek, other people's maven/repo is dark magic, and those are not super helpful errros. Might well be the weekend before I can take a look...

bollwyvl commented on recipe/bld.bat:
```diff
@@ -11,5 +11,9 @@ cmd.exe /c mvn --batch-mode --projects=tool clean || echo ""
 echo IF %%ERRORLEVEL%% NEQ 0 EXIT /B %%ERRORLEVEL%% >> %LIBRARY_BIN%\antlr4.cmd
+
+echo java -Xmx500M -cp %LIBRARY_LIB%\\antlr4-%PKG_VERSION%-complete.jar;%%CLASSPATH%% org.antlr.v4.gui.TestRig %%* > %LIBRARY_BIN%\grun.cmd
+echo IF %%ERRORLEVEL%% NEQ 0 EXIT /B %%ERRORLEVEL%% >> %LIBRARY_BIN%\grun.cmd
```
:+1:

Commit:
Try to fix maven build

Commit:
More maven attempts

Commit:
One more maven try

Commit:
Another poke at maven and antlr4

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found some lint.

Here's what I've got...


For **recipe**:

 * Failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint .` from the recipe directory.

Commit:
Fix meta.yaml

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

Commit:
Force maven version

Commit:
Force maven version also in Windows bld.bat

zklaus commented:
Sooooo I did some digging and it seems there is mainly one problem: The versioning by the upstream project.

The release tarball for 4.8 uses a different version in the `pom.xml`, namely `4.8-2-SNAPSHOT`. At previous points in the history the versions `4.8-SNAPSHOT`, `4.8`, `4.9-SNAPSHOT`, `4.8-1-SNAPSHOT`, `4.8-1`, `4.8-2-SNAPSHOT`. All of this happened over the period of three days in January 2020, so I guess there was just some confusion. I am now wondering if maybe the changed checksum is also due to an actually changed content.

Anyway, one solution that seems to work is to force the version `4.8` using `mvn versions:set -DnewVersion=4.8` before the actual maven build. This seems to be in line with the intention of upstream. Likely this should be revisited for the next release.

What do you think?

bollwyvl commented on recipe/build.sh:
```diff
@@ -1,18 +1,18 @@
 
-mvn --batch-mode --projects=tool clean
-mvn --batch-mode --projects=tool -DskipTests install
+mvn --batch-mode versions:set -DnewVersion=${PKG_VERSION}
```
:+1:

bollwyvl commented on recipe/meta.yaml:
```diff
@@ -1,17 +1,19 @@
+{% set major_version = "4" %}
+{% set minor_version = "8" %}
+{% set version = major_version + "." + minor_version %}
+{% set revision = "1" %}
```
anyhow is perhaps a bridge too far... the bot _really_ likes there to be a jinja var called `version` which is a string, so that it can try updating it when the upstream changes. Also, there have historically been `x.y.z` releases, so this would need to be fixed again, again...

zklaus replied:
Yeah, agreed. This is mostly a remnant of me trying to figure things out before understanding the maven issue fully. Will clean up.

bollwyvl commented on recipe/meta.yaml:
```diff
@@ -1,17 +1,19 @@
-  url: https://github.com/{{ name }}/{{ name }}{{ version[0] }}/archive/{{ version }}.tar.gz
-  sha256: c8e41fa04e4d79ec2dc91dad9ddfd202b09b50fd59f17f3a10f52dbdffdaf70d
+  url: https://github.com/{{ name }}/{{ name }}{{ major_version }}/archive/{{ version }}.tar.gz
+  sha256: 992d52444b81ed75e52ea62f9f38ecb7652d5ce2a2130af143912b3042a6d77e
```
yeah, i'm grumpy about the sha changes...

zklaus replied:
Me too...

bollwyvl commented on recipe/meta.yaml:
```diff
@@ -22,6 +24,7 @@ requirements:
 test:
   commands:
     - antlr4
+    - grun
```
:+1:

bollwyvl commented:
Good job on finding the maven version flag _and_ using `PKG_VERSION` to reference it.

Modulo the comment about keeping `version` bot-friendly (and resilient to, e.g. `4.8.1`) this is looking good!

zklaus commented:
Thanks! Looking at [the official antlr maven repo](https://repo1.maven.org/maven2/org/antlr/antlr4/), it seems like really the released version should be something like 4.8-1, but conda didn't like the hyphen in the version number and I was not convinced by the tags and commits in the antlr repo.

Do you think going with just 4.8 is ok?

Should we somehow place a warning in the recipe to check on the next release?

Commit:
Simplified meta.yaml

bollwyvl commented:
Yeah, i really wish they put their source distros out on that release site, with hash sums, rather than relying on things on github which can (apparently) change.

conda indeed will not like having a `-` separator in the version itself... it looks like we _could_ use `_`, however:

https://github.com/conda/conda/blob/305bf88eb8ce6d79dab8888ef34e93e72e35d822/conda/models/version.py#L102

So i guess we should try to release this as version `4.8_1`, build `number: 0` (as that's for us, in case we mess something up). Since it isn't part of the tag, I don't really know how we can automate this in the future... I'm watching release there, that might be the best we can do...

Commit:
Update version to include final tag (original 4.8-1, coded as 4.8_1)

zklaus commented:
Ok, I followed your suggestion and put "4.8_1" as version into the `meta.yaml`. I was unsure whether to change that back to "4.8-1" for the maven file, but ultimately decided to stick with 4.8_1 also there, partly because I didn't know how to do that in the Windows build (`${PKG_VERSION/_/-}` is easy enough in Linux), partly because I like the increased consistency. Ok like this?

bollwyvl committed:
tweak how version string is captured/mangled with patch level

bollwyvl commented:
@zklaus thanks for sticking with it! i think we're getting very close! I'll make a sub-PR with a minor change to how the version is captured... it's not going to help us with the bot much, but I fear we'll forget what we've done here.

To that end: I really appreciate the digging you did on maven, which I'll confess has been a somewhat formidable challenge to me at times, as I'm not a frequent user! Are you interested in co-maintaining? If so, feel free to add your name to the `feedstock-maintainers` key on your PR. Or not, if you're not looking for another hobby!

Commit:
Merge pull request #1 from bollwyvl/add-grun-script

tweak how version string is captured/mangled with patch level

Commit:
Add zklaus to maintainers

zklaus commented:
@bollwyvl, thanks for the PR. This all seems reasonable. I did add myself to the maintainers, though I don't think of myself as a maven expert :)

bollwyvl commented:
thanks again!

zklaus commented:
Cheers, my pleasure :smile:
